gnovm/pkg/gnolang/preprocess.go

```diff
@@ -3673,6 +3912,172 @@ func findDependentNames(n Node, dst map[Name]struct{}) {
 	}
 }
 
+func isBlockNodeInLoop(store Store, bn BlockNode) bool {
+	for bn != nil {
+		bn = bn.GetParentNode(store)
+		switch bn.(type) {
+		case *FuncLitExpr:
+			return false
+		case *ForStmt:
+			return true
+		case *RangeStmt:
+			return true
+		default:
+			continue
+		}
+	}
+	return false
+}
+
+// reset value path, static block, preprocessed attr, etc.
+func resetStaticBlock(bn BlockNode) {
+	Transcribe(bn, func(ns []Node, ftype TransField, index int, n Node, stage TransStage) (Node, TransCtrl) {
+		switch stage {
+		case TRANS_ENTER:
+			switch cn := n.(type) {
+			case *NameExpr:
+				cn.Path = NewValuePath(VPUverse, 0, 0, cn.Name)
+				cn.SetAttribute(ATTR_PREPROCESSED, false)
+				return cn, TRANS_CONTINUE
+			case BlockNode:
+				cn.GetStaticBlock().reset()
+				cn.SetAttribute(ATTR_PREPROCESSED, false)
+				return cn, TRANS_CONTINUE
+			default:
+				cn.SetAttribute(ATTR_PREPROCESSED, false)
+				return cn, TRANS_CONTINUE
+			}
+		default:
+			return n, TRANS_CONTINUE
+		}
+	})
+}
+
+// for goto...
+// wrap goto loop into block stmt.
+func rebuildBody(b Body, gloop *LoopInfo, loc Location) Body {
```

Comment:
unfortunately, i don't think this works for the following kind of goto-loop.

```go
LABEL1:
...
LABEL2:
...
if cond1 {
    goto LABEL1
}
...
if cond2 {
    goto LABEL2
}
```

notice how goto LABEL1 and goto LABEL2 make two loops that intersect, but one loop is not nested within the other.